src/main/java/org/broadinstitute/hellbender/tools/walkers/qc/TransferReadTags.java

```diff
@@ -0,0 +1,163 @@
+        alignedSamIterator = new PeekableIterator<>(directlyAccessEngineReadsDataSource().iterator());
+        final SAMFileHeader.SortOrder sortOrderAlignedReads = directlyAccessEngineReadsDataSource().getHeader().getSortOrder();
+        Utils.validate(sortOrderAlignedReads == SAMFileHeader.SortOrder.queryname, "aligned sam must be sorted by queryname");
+
+        unmappedSam = new ReadsPathDataSource(unmappedSamFile.toPath());
+
+        // We would check that the unmapped bam is query-name sorted, but
+        // the SortOrder field is often not populated. Thus we simply assume that
+        // the unmapped sam is query-name sorted and rely on the traversal to throw an
+        // error when it is not.
+        unmappedSamIterator = new PeekableIterator<>(unmappedSam.iterator());
+
+        // Initialize the current unmapped read. The counterpart for the aligned read will be initialized in traverse()
+        if (unmappedSamIterator.hasNext()){
+            currentUnmappedRead = unmappedSamIterator.next();
+        } else {
+            throw new UserException("unmapped sam iterator is empty.");
+        }
+
+        writer = createSAMWriter(new GATKPath(outSamFile.getAbsolutePath()), false);
+    }
+
+    /**
+     * The traversal assumes that the aligned (target) reads is a subset of
+     * the unmapped reads. The tool exists when there exists a read in the
+     * aligned read that is not present in the unmapped read.
+     */
+    @Override
+    public void traverse() {
+        // will need to use this later
+        while (alignedSamIterator.hasNext()){
+            currentTargetRead = alignedSamIterator.next();
+            int diff = queryNameComparator.compareReadNames(currentTargetRead, currentUnmappedRead);
+
+            if (diff == 0) {
+                // The query names match.
+                final GATKRead updatedRead = updateReadTags(currentTargetRead, currentUnmappedRead);
+                writer.addRead(updatedRead);
+                progressMeter.update(currentTargetRead);
+                continue;
+            } else if (diff > 0){
+                // target read is ahead; play unmapped reads forward until it catches up.
+                while (unmappedSamIterator.hasNext()){
+                    currentUnmappedRead = unmappedSamIterator.next();
+                    diff = queryNameComparator.compareReadNames(currentTargetRead, currentUnmappedRead);
+                    if (diff > 0){
+                        continue;
+                    } else if (diff == 0){
+                        // caught up: start moving the aligned reads forward
+                        final GATKRead updatedRead = updateReadTags(currentTargetRead, currentUnmappedRead);
+                        writer.addRead(updatedRead);
+                        break;
+                    } else {
+                        throw new IllegalStateException("Aligned read is lexicographically smaller than the unmapped read. This tool assumes " +
```

Comment:
I would say the regular case here is that a read is missing from the unmapped bam file and I wold probably make that the primary takeaway (with a note that thsi could also be caused by sorting woes).

Reply:
Reworded